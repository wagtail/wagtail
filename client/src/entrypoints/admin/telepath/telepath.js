import Telepath from 'telepath-unpack';

// Prevent double-initialisation if imported twice
// This was introduced because both the sidebar and the page editor
// use telepath, but they use separate hooks for JS dependencies so we
// can't de-duplicate the telepath import.
if (!window.telepath) {
  window.telepath = new Telepath();
}
