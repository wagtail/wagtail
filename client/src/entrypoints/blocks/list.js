import $ from 'jquery';

// eslint-disable-next-line func-names
window.ListBlock = function (opts) {
  /* contents of 'opts':
          definitionPrefix (required)
          childInitializer (optional) - JS initializer function for each child
      */
  const listMemberTemplate = $('#' + opts.definitionPrefix + '-newmember').text();

  // eslint-disable-next-line func-names
  return function (elementPrefix) {
    // eslint-disable-next-line no-undef, new-cap
    const sequence = Sequence({
      prefix: elementPrefix,
      maxNumChildBlocks: Infinity,
      onInitializeMember(sequenceMember) {
        /* initialize child block's JS behaviour */
        if (opts.childInitializer) {
          opts.childInitializer(sequenceMember.prefix + '-value');
        }

        /* initialise delete button */
        $('#' + sequenceMember.prefix + '-delete').on('click', () => {
          sequenceMember.delete();
        });

        /* initialise move up/down buttons */
        $('#' + sequenceMember.prefix + '-moveup').on('click', () => {
          sequenceMember.moveUp();
        });

        $('#' + sequenceMember.prefix + '-movedown').on('click', () => {
          sequenceMember.moveDown();
        });
      },

      onEnableMoveUp(sequenceMember) {
        $('#' + sequenceMember.prefix + '-moveup').removeClass('disabled');
      },

      onDisableMoveUp(sequenceMember) {
        $('#' + sequenceMember.prefix + '-moveup').addClass('disabled');
      },

      onEnableMoveDown(sequenceMember) {
        $('#' + sequenceMember.prefix + '-movedown').removeClass('disabled');
      },

      onDisableMoveDown(sequenceMember) {
        $('#' + sequenceMember.prefix + '-movedown').addClass('disabled');
      }
    });

    /* initialize 'add' button */
    $('#' + elementPrefix + '-add').on('click', () => {
      sequence.insertMemberAtEnd(listMemberTemplate);
    });
  };
};
