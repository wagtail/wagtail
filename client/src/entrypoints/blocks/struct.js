// eslint-disable-next-line func-names
window.StructBlock = function (childInitializersByName) {
  // eslint-disable-next-line func-names
  return function (prefix) {
    // eslint-disable-next-line no-restricted-syntax, guard-for-in
    for (const childName in childInitializersByName) {
      const childInitializer = childInitializersByName[childName];
      const childPrefix = prefix + '-' + childName;
      childInitializer(childPrefix);
    }
  };
};
