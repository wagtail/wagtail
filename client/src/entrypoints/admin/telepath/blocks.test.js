/* eslint-disable no-unused-vars */
import './telepath';
import './blocks';

import $ from 'jquery';
window.$ = $;

// Define some callbacks in global scope that can be mocked in tests
let constructor = (_widgetName, _name, _id, _initialState) => {};
let setState = (_widgetName, _state) => {};
let getState = (_widgetName) => {};
let getValue = (_widgetName) => {};
let focus = (_widgetName) => {};

class DummyWidget {
  constructor(widgetName) {
    this.widgetName = widgetName;
  }

  render(placeholder, name, id, initialState) {
    const widgetName = this.widgetName;
    constructor(widgetName, { name, id, initialState });

    $(placeholder).replaceWith(`<p name="${name}" id="${id}">${widgetName}</p>`);
    return {
      setState(state) { setState(widgetName, state); },
      getState() { getState(widgetName); return `state: ${widgetName}`; },
      getValue() { getValue(widgetName); return `value: ${widgetName}`; },
      focus() { focus(widgetName); },
    };
  }
}
window.telepath.register('wagtail.widgets.DummyWidget', DummyWidget);

describe('telepath: wagtail.blocks.FieldBlock', () => {
  let boundField;

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Create a placeholder to render the block
    document.body.innerHTML = '<div id="placeholder"></div>';

    // Unpack and render
    const fieldDef = window.telepath.unpack({
      _type: 'wagtail.blocks.FieldBlock',
      _args: ['test_field', {
        _type: 'wagtail.widgets.DummyWidget',
        _args: ['The widget']
      }, {
        label: 'Test Field',
        required: true,
        icon: 'placeholder',
        classname: 'field char_field widget-text_input fieldname-test_charblock'
      }]
    });
    boundField = fieldDef.render($('#placeholder'), 'the-prefix', 'Test initial state');
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('constructor is called with correct parameters', () => {
    expect(constructor.mock.calls.length).toBe(1);
    expect(constructor.mock.calls[0][0]).toBe('The widget');
    expect(constructor.mock.calls[0][1]).toEqual({
      name: 'the-prefix',
      id: 'the-prefix',
      initialState: 'Test initial state',
    });
  });

  test('getValue() calls widget getValue()', () => {
    const value = boundField.getValue();
    expect(getValue.mock.calls.length).toBe(1);
    expect(value).toEqual('value: The widget');
  });

  test('getState() calls widget getState()', () => {
    const state = boundField.getState();
    expect(getState.mock.calls.length).toBe(1);
    expect(state).toEqual('state: The widget');
  });

  test('setState() calls widget setState()', () => {
    boundField.setState('Test changed state');
    expect(setState.mock.calls.length).toBe(1);
    expect(setState.mock.calls[0][0]).toBe('The widget');
    expect(setState.mock.calls[0][1]).toBe('Test changed state');
  });

  test('focus() calls widget focus()', () => {
    boundField.focus();
    expect(focus.mock.calls.length).toBe(1);
    expect(focus.mock.calls[0][0]).toBe('The widget');
  });
});

describe('telepath: wagtail.blocks.StructBlock', () => {
  let boundField;

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Create a placeholder to render the block
    document.body.innerHTML = '<div id="placeholder"></div>';

    // Unpack and render
    const fieldDef = window.telepath.unpack({
      _type: 'wagtail.blocks.StructBlock',
      _args: ['heading_block', [{
        _type: 'wagtail.blocks.FieldBlock',
        _args: ['heading_text', {
          _type: 'wagtail.widgets.DummyWidget',
          _args: ['Heading widget']
        }, {
          label: 'Heading text',
          required: true,
          icon: 'placeholder',
          classname: 'field char_field widget-text_input fieldname-heading_text'
        }]
      }, {
        _type: 'wagtail.blocks.FieldBlock',
        _args: ['size', {
          _type: 'wagtail.widgets.DummyWidget',
          _args: ['Size widget']
        }, {
          label: 'Size',
          required: false,
          icon: 'placeholder',
          classname: 'field choice_field widget-select fieldname-size'
        }]
      }], {
        label: 'Heading block',
        required: false,
        icon: 'title',
        classname: 'struct-block',
        helpText: null
      }]
    });
    boundField = fieldDef.render($('#placeholder'), 'the-prefix', {
      heading_text: 'Test heading text',
      size: '123'
    });
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('Widget constructors are called with correct parameters', () => {
    expect(constructor.mock.calls.length).toBe(2);

    expect(constructor.mock.calls[0][0]).toBe('Heading widget');
    expect(constructor.mock.calls[0][1]).toEqual({
      name: 'the-prefix-heading_text',
      id: 'the-prefix-heading_text',
      initialState: 'Test heading text',
    });

    expect(constructor.mock.calls[1][0]).toBe('Size widget');
    expect(constructor.mock.calls[1][1]).toEqual({
      name: 'the-prefix-size',
      id: 'the-prefix-size',
      initialState: '123',
    });
  });

  test('getValue() calls getValue() on all widgets', () => {
    const value = boundField.getValue();
    expect(getValue.mock.calls.length).toBe(2);
    expect(value).toEqual({
      heading_text: 'value: Heading widget',
      size: 'value: Size widget'
    });
  });

  test('getState() calls getState() on all widgets', () => {
    const state = boundField.getState();
    expect(getState.mock.calls.length).toBe(2);
    expect(state).toEqual({
      heading_text: 'state: Heading widget',
      size: 'state: Size widget'
    });
  });

  test('setState() calls setState() on all widgets', () => {
    boundField.setState({
      heading_text: 'Changed heading text',
      size: '456'
    });
    expect(setState.mock.calls.length).toBe(2);
    expect(setState.mock.calls[0][0]).toBe('Heading widget');
    expect(setState.mock.calls[0][1]).toBe('Changed heading text');
    expect(setState.mock.calls[1][0]).toBe('Size widget');
    expect(setState.mock.calls[1][1]).toBe('456');
  });

  test('focus() calls focus() on first widget', () => {
    boundField.focus();
    expect(focus.mock.calls.length).toBe(1);
    expect(focus.mock.calls[0][0]).toBe('Heading widget');
  });
});

describe('telepath: wagtail.blocks.ListBlock', () => {
  let boundField;

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Create a placeholder to render the block
    document.body.innerHTML = '<div id="placeholder"></div>';

    // Unpack and render
    const fieldDef = window.telepath.unpack({
      _type: 'wagtail.blocks.ListBlock',
      _args: ['test_listblock', {
        _type: 'wagtail.blocks.FieldBlock',
        _args: ['', {
          _type: 'wagtail.widgets.DummyWidget',
          _args: ['The widget']
        }, {
          label: '',
          required: true,
          icon: 'pilcrow',
          classname: 'field char_field widget-admin_auto_height_text_input fieldname-'
        }]
      }, null, {
        label: 'Test listblock',
        icon: 'placeholder',
        classname: null,
        helpText: null
      }]
    });
    boundField = fieldDef.render($('#placeholder'), 'the-prefix', [
      'First value',
      'Second value'
    ]);
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('Widget constructors are called with correct parameters', () => {
    expect(constructor.mock.calls.length).toBe(2);

    expect(constructor.mock.calls[0][0]).toBe('The widget');
    expect(constructor.mock.calls[0][1]).toEqual({
      name: 'the-prefix-0-value',
      id: 'the-prefix-0-value',
      initialState: 'First value',
    });

    expect(constructor.mock.calls[1][0]).toBe('The widget');
    expect(constructor.mock.calls[1][1]).toEqual({
      name: 'the-prefix-1-value',
      id: 'the-prefix-1-value',
      initialState: 'Second value',
    });
  });

  test('getValue() calls getValue() on widget for both values', () => {
    const value = boundField.getValue();
    expect(getValue.mock.calls.length).toBe(2);
    expect(value).toEqual([
      'value: The widget',
      'value: The widget'
    ]);
  });

  test('getState() calls getState() on all widgets', () => {
    const state = boundField.getState();
    expect(getState.mock.calls.length).toBe(2);
    expect(state).toEqual([
      'state: The widget',
      'state: The widget'
    ]);
  });

  test('setState() creates new widgets', () => {
    boundField.setState([
      'Changed first value',
      'Changed second value',
      'Third value'
    ]);

    // Includes the two initial calls, plus the three new ones
    expect(constructor.mock.calls.length).toBe(5);

    expect(constructor.mock.calls[2][0]).toBe('The widget');
    expect(constructor.mock.calls[2][1]).toEqual({
      name: 'the-prefix-0-value',
      id: 'the-prefix-0-value',
      initialState: 'Changed first value',
    });

    expect(constructor.mock.calls[3][0]).toBe('The widget');
    expect(constructor.mock.calls[3][1]).toEqual({
      name: 'the-prefix-1-value',
      id: 'the-prefix-1-value',
      initialState: 'Changed second value',
    });

    expect(constructor.mock.calls[4][0]).toBe('The widget');
    expect(constructor.mock.calls[4][1]).toEqual({
      name: 'the-prefix-2-value',
      id: 'the-prefix-2-value',
      initialState: 'Third value',
    });

    // Let's get the state now to make sure the initial widgets are gone
    const state = boundField.getState();
    expect(getState.mock.calls.length).toBe(3);
    expect(state).toEqual([
      'state: The widget',
      'state: The widget',
      'state: The widget'
    ]);
  });

  test('focus() calls focus() on first widget', () => {
    boundField.focus();
    expect(focus.mock.calls.length).toBe(1);
    expect(focus.mock.calls[0][0]).toBe('The widget');
  });
});
