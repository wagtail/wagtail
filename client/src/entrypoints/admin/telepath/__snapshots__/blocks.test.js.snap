// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telepath: wagtail.blocks.FieldBlock it renders correctly 1`] = `
"<div class=\\"field char_field widget-text_input fieldname-test_charblock\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix\\" id=\\"the-prefix\\">The widget</p>
            <span></span>
          </div>
        <p class=\\"help\\">drink <em>more</em> water</p></div>
      </div>"
`;

exports[`telepath: wagtail.blocks.ListBlock deleteBlock() deletes a block 1`] = `
"<span>
          <div class=\\"help\\">
            <div class=\\"icon-help\\">?</div>
            use <strong>a few</strong> of these
          </div>
        </span><div class=\\"c-sf-container \\">
        <input type=\\"hidden\\" name=\\"the-prefix-count\\" data-streamfield-list-count=\\"\\" value=\\"2\\">

        <div data-streamfield-list-container=\\"\\"><div id=\\"the-prefix-0-container\\" aria-hidden=\\"false\\">
        <input type=\\"hidden\\" data-listblock-deleted=\\"\\" id=\\"the-prefix-0-deleted\\" name=\\"the-prefix-0-deleted\\" value=\\"\\">
        <input type=\\"hidden\\" data-listblock-index=\\"\\" id=\\"the-prefix-0-order\\" name=\\"the-prefix-0-order\\" value=\\"0\\">
        <div>
          <div class=\\"c-sf-container__block-container\\">
            <div class=\\"c-sf-block\\">
              <div class=\\"c-sf-block__header\\">
                <span class=\\"c-sf-block__header__icon\\">
                  <i class=\\"icon icon-pilcrow\\"></i>
                </span>
                <h3 class=\\"c-sf-block__header__title\\"></h3>
                <div class=\\"c-sf-block__actions\\">
                  <span class=\\"c-sf-block__type\\"></span>
                  <button type=\\"button\\" id=\\"the-prefix-0-moveup\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move up' %}\\">
                    <i class=\\"icon icon-arrow-up\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" id=\\"the-prefix-0-movedown\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move down' %}\\">
                    <i class=\\"icon icon-arrow-down\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" data-listblock-delete-button=\\"\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Delete' %}\\">
                    <i class=\\"icon icon-bin\\" aria-hidden=\\"true\\"></i>
                  </button>
                </div>
              </div>
              <div class=\\"c-sf-block__content\\" aria-hidden=\\"false\\">
                <div class=\\"c-sf-block__content-inner\\">
                  <div class=\\"field char_field widget-admin_auto_height_text_input fieldname-\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-0-value\\" id=\\"the-prefix-0-value\\">The widget</p>
            <span></span>
          </div>
        </div>
      </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div><div id=\\"the-prefix-1-container\\" aria-hidden=\\"true\\" style=\\"overflow: hidden; height: 0px; padding-top: 0px; margin-top: 0px; padding-bottom: 0px; margin-bottom: 0px;\\">
        <input type=\\"hidden\\" data-listblock-deleted=\\"\\" id=\\"the-prefix-1-deleted\\" name=\\"the-prefix-1-deleted\\" value=\\"1\\">
        <input type=\\"hidden\\" data-listblock-index=\\"\\" id=\\"the-prefix-1-order\\" name=\\"the-prefix-1-order\\" value=\\"1\\">
        <div>
          <div class=\\"c-sf-container__block-container\\">
            <div class=\\"c-sf-block\\">
              <div class=\\"c-sf-block__header\\">
                <span class=\\"c-sf-block__header__icon\\">
                  <i class=\\"icon icon-pilcrow\\"></i>
                </span>
                <h3 class=\\"c-sf-block__header__title\\"></h3>
                <div class=\\"c-sf-block__actions\\">
                  <span class=\\"c-sf-block__type\\"></span>
                  <button type=\\"button\\" id=\\"the-prefix-1-moveup\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move up' %}\\">
                    <i class=\\"icon icon-arrow-up\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" id=\\"the-prefix-1-movedown\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move down' %}\\">
                    <i class=\\"icon icon-arrow-down\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" data-listblock-delete-button=\\"\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Delete' %}\\">
                    <i class=\\"icon icon-bin\\" aria-hidden=\\"true\\"></i>
                  </button>
                </div>
              </div>
              <div class=\\"c-sf-block__content\\" aria-hidden=\\"false\\">
                <div class=\\"c-sf-block__content-inner\\">
                  <div class=\\"field char_field widget-admin_auto_height_text_input fieldname-\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-1-value\\" id=\\"the-prefix-1-value\\">The widget</p>
            <span></span>
          </div>
        </div>
      </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div></div>
        <button type=\\"button\\" title=\\"Add\\" data-streamfield-list-add=\\"\\" class=\\"c-sf-add-button c-sf-add-button--visible\\">
          <i aria-hidden=\\"true\\">+</i>
        </button>
      </div>"
`;

exports[`telepath: wagtail.blocks.ListBlock it renders correctly 1`] = `
"<span>
          <div class=\\"help\\">
            <div class=\\"icon-help\\">?</div>
            use <strong>a few</strong> of these
          </div>
        </span><div class=\\"c-sf-container \\">
        <input type=\\"hidden\\" name=\\"the-prefix-count\\" data-streamfield-list-count=\\"\\" value=\\"2\\">

        <div data-streamfield-list-container=\\"\\"><div id=\\"the-prefix-0-container\\" aria-hidden=\\"false\\">
        <input type=\\"hidden\\" data-listblock-deleted=\\"\\" id=\\"the-prefix-0-deleted\\" name=\\"the-prefix-0-deleted\\" value=\\"\\">
        <input type=\\"hidden\\" data-listblock-index=\\"\\" id=\\"the-prefix-0-order\\" name=\\"the-prefix-0-order\\" value=\\"0\\">
        <div>
          <div class=\\"c-sf-container__block-container\\">
            <div class=\\"c-sf-block\\">
              <div class=\\"c-sf-block__header\\">
                <span class=\\"c-sf-block__header__icon\\">
                  <i class=\\"icon icon-pilcrow\\"></i>
                </span>
                <h3 class=\\"c-sf-block__header__title\\"></h3>
                <div class=\\"c-sf-block__actions\\">
                  <span class=\\"c-sf-block__type\\"></span>
                  <button type=\\"button\\" id=\\"the-prefix-0-moveup\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move up' %}\\">
                    <i class=\\"icon icon-arrow-up\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" id=\\"the-prefix-0-movedown\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move down' %}\\">
                    <i class=\\"icon icon-arrow-down\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" data-listblock-delete-button=\\"\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Delete' %}\\">
                    <i class=\\"icon icon-bin\\" aria-hidden=\\"true\\"></i>
                  </button>
                </div>
              </div>
              <div class=\\"c-sf-block__content\\" aria-hidden=\\"false\\">
                <div class=\\"c-sf-block__content-inner\\">
                  <div class=\\"field char_field widget-admin_auto_height_text_input fieldname-\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-0-value\\" id=\\"the-prefix-0-value\\">The widget</p>
            <span></span>
          </div>
        </div>
      </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div><div id=\\"the-prefix-1-container\\" aria-hidden=\\"false\\">
        <input type=\\"hidden\\" data-listblock-deleted=\\"\\" id=\\"the-prefix-1-deleted\\" name=\\"the-prefix-1-deleted\\" value=\\"\\">
        <input type=\\"hidden\\" data-listblock-index=\\"\\" id=\\"the-prefix-1-order\\" name=\\"the-prefix-1-order\\" value=\\"1\\">
        <div>
          <div class=\\"c-sf-container__block-container\\">
            <div class=\\"c-sf-block\\">
              <div class=\\"c-sf-block__header\\">
                <span class=\\"c-sf-block__header__icon\\">
                  <i class=\\"icon icon-pilcrow\\"></i>
                </span>
                <h3 class=\\"c-sf-block__header__title\\"></h3>
                <div class=\\"c-sf-block__actions\\">
                  <span class=\\"c-sf-block__type\\"></span>
                  <button type=\\"button\\" id=\\"the-prefix-1-moveup\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move up' %}\\">
                    <i class=\\"icon icon-arrow-up\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" id=\\"the-prefix-1-movedown\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move down' %}\\">
                    <i class=\\"icon icon-arrow-down\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" data-listblock-delete-button=\\"\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Delete' %}\\">
                    <i class=\\"icon icon-bin\\" aria-hidden=\\"true\\"></i>
                  </button>
                </div>
              </div>
              <div class=\\"c-sf-block__content\\" aria-hidden=\\"false\\">
                <div class=\\"c-sf-block__content-inner\\">
                  <div class=\\"field char_field widget-admin_auto_height_text_input fieldname-\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-1-value\\" id=\\"the-prefix-1-value\\">The widget</p>
            <span></span>
          </div>
        </div>
      </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div></div>
        <button type=\\"button\\" title=\\"Add\\" data-streamfield-list-add=\\"\\" class=\\"c-sf-add-button c-sf-add-button--visible\\">
          <i aria-hidden=\\"true\\">+</i>
        </button>
      </div>"
`;

exports[`telepath: wagtail.blocks.StreamBlock it renders correctly 1`] = `
"<span>
          <div class=\\"help\\">
            <div class=\\"icon-help\\">?</div>
            use <strong>plenty</strong> of these
          </div>
        </span><div class=\\"c-sf-container \\">
        <input type=\\"hidden\\" name=\\"the-prefix-count\\" data-streamfield-stream-count=\\"\\" value=\\"2\\">
        <div data-streamfield-stream-container=\\"\\"><div>
        <button data-streamblock-menu-open=\\"\\" type=\\"button\\" title=\\"Add\\" class=\\"c-sf-add-button c-sf-add-button--visible\\">
          <i aria-hidden=\\"true\\">+</i>
        </button>
        <div data-streamblock-menu-outer=\\"\\" style=\\"display: none;\\" aria-hidden=\\"true\\">
          <div data-streamblock-menu-inner=\\"\\" class=\\"c-sf-add-panel\\"><div class=\\"c-sf-add-panel__grid\\"><button type=\\"button\\" class=\\"c-sf-button action-add-block-test_block_a\\">
            <span class=\\"c-sf-button__icon\\">
              <i class=\\"icon icon-placeholder\\"></i>
            </span>
            <span class=\\"c-sf-button__label\\">Test Block B</span>
          </button><button type=\\"button\\" class=\\"c-sf-button action-add-block-test_block_b\\">
            <span class=\\"c-sf-button__icon\\">
              <i class=\\"icon icon-pilcrow\\"></i>
            </span>
            <span class=\\"c-sf-button__label\\">Test Block B</span>
          </button></div></div>
        </div>
      </div><div aria-hidden=\\"false\\">
        <input type=\\"hidden\\" data-streamblock-deleted=\\"\\" name=\\"the-prefix-0-deleted\\" value=\\"\\">
        <input type=\\"hidden\\" data-streamblock-index=\\"\\" name=\\"the-prefix-0-order\\" value=\\"0\\">
        <input type=\\"hidden\\" name=\\"the-prefix-0-type\\" value=\\"test_block_a\\">
        <input type=\\"hidden\\" name=\\"the-prefix-0-id\\" value=\\"1\\">

        <div>
          <div class=\\"c-sf-container__block-container\\">
            <div class=\\"c-sf-block\\">
              <div class=\\"c-sf-block__header\\">
                <span class=\\"c-sf-block__header__icon\\">
                  <i class=\\"icon icon-placeholder\\"></i>
                </span>
                <h3 class=\\"c-sf-block__header__title\\"></h3>
                <div class=\\"c-sf-block__actions\\">
                  <span class=\\"c-sf-block__type\\">Test Block B</span>
                  <button type=\\"button\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move up' %}\\">
                    <i class=\\"icon icon-arrow-up\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move down' %}\\">
                    <i class=\\"icon icon-arrow-down\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" data-streamblock-delete-button=\\"\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Delete' %}\\">
                    <i class=\\"icon icon-bin\\" aria-hidden=\\"true\\"></i>
                  </button>
                </div>
              </div>
              <div class=\\"c-sf-block__content\\" aria-hidden=\\"false\\">
                <div class=\\"c-sf-block__content-inner\\">
                  <div class=\\"field char_field widget-text_input fieldname-test_charblock\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-0-value\\" id=\\"the-prefix-0-value\\">Block A widget</p>
            <span></span>
          </div>
        </div>
      </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div><div>
        <button data-streamblock-menu-open=\\"\\" type=\\"button\\" title=\\"Add\\" class=\\"c-sf-add-button c-sf-add-button--visible\\">
          <i aria-hidden=\\"true\\">+</i>
        </button>
        <div data-streamblock-menu-outer=\\"\\" style=\\"display: none;\\" aria-hidden=\\"true\\">
          <div data-streamblock-menu-inner=\\"\\" class=\\"c-sf-add-panel\\"><div class=\\"c-sf-add-panel__grid\\"><button type=\\"button\\" class=\\"c-sf-button action-add-block-test_block_a\\">
            <span class=\\"c-sf-button__icon\\">
              <i class=\\"icon icon-placeholder\\"></i>
            </span>
            <span class=\\"c-sf-button__label\\">Test Block B</span>
          </button><button type=\\"button\\" class=\\"c-sf-button action-add-block-test_block_b\\">
            <span class=\\"c-sf-button__icon\\">
              <i class=\\"icon icon-pilcrow\\"></i>
            </span>
            <span class=\\"c-sf-button__label\\">Test Block B</span>
          </button></div></div>
        </div>
      </div><div aria-hidden=\\"false\\">
        <input type=\\"hidden\\" data-streamblock-deleted=\\"\\" name=\\"the-prefix-1-deleted\\" value=\\"\\">
        <input type=\\"hidden\\" data-streamblock-index=\\"\\" name=\\"the-prefix-1-order\\" value=\\"1\\">
        <input type=\\"hidden\\" name=\\"the-prefix-1-type\\" value=\\"test_block_b\\">
        <input type=\\"hidden\\" name=\\"the-prefix-1-id\\" value=\\"2\\">

        <div>
          <div class=\\"c-sf-container__block-container\\">
            <div class=\\"c-sf-block\\">
              <div class=\\"c-sf-block__header\\">
                <span class=\\"c-sf-block__header__icon\\">
                  <i class=\\"icon icon-pilcrow\\"></i>
                </span>
                <h3 class=\\"c-sf-block__header__title\\"></h3>
                <div class=\\"c-sf-block__actions\\">
                  <span class=\\"c-sf-block__type\\">Test Block B</span>
                  <button type=\\"button\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move up' %}\\">
                    <i class=\\"icon icon-arrow-up\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Move down' %}\\">
                    <i class=\\"icon icon-arrow-down\\" aria-hidden=\\"true\\"></i>
                  </button>
                  <button type=\\"button\\" data-streamblock-delete-button=\\"\\" class=\\"c-sf-block__actions__single\\" title=\\"{% trans 'Delete' %}\\">
                    <i class=\\"icon icon-bin\\" aria-hidden=\\"true\\"></i>
                  </button>
                </div>
              </div>
              <div class=\\"c-sf-block__content\\" aria-hidden=\\"false\\">
                <div class=\\"c-sf-block__content-inner\\">
                  <div class=\\"field char_field widget-admin_auto_height_text_input fieldname-test_textblock\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-1-value\\" id=\\"the-prefix-1-value\\">Block B widget</p>
            <span></span>
          </div>
        </div>
      </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div><div>
        <button data-streamblock-menu-open=\\"\\" type=\\"button\\" title=\\"Add\\" class=\\"c-sf-add-button c-sf-add-button--visible\\">
          <i aria-hidden=\\"true\\">+</i>
        </button>
        <div data-streamblock-menu-outer=\\"\\" style=\\"display: none;\\" aria-hidden=\\"true\\">
          <div data-streamblock-menu-inner=\\"\\" class=\\"c-sf-add-panel\\"><div class=\\"c-sf-add-panel__grid\\"><button type=\\"button\\" class=\\"c-sf-button action-add-block-test_block_a\\">
            <span class=\\"c-sf-button__icon\\">
              <i class=\\"icon icon-placeholder\\"></i>
            </span>
            <span class=\\"c-sf-button__label\\">Test Block B</span>
          </button><button type=\\"button\\" class=\\"c-sf-button action-add-block-test_block_b\\">
            <span class=\\"c-sf-button__icon\\">
              <i class=\\"icon icon-pilcrow\\"></i>
            </span>
            <span class=\\"c-sf-button__label\\">Test Block B</span>
          </button></div></div>
        </div>
      </div></div>
      </div>"
`;

exports[`telepath: wagtail.blocks.StructBlock it renders correctly 1`] = `
"<div class=\\"struct-block\\">
        
          <span>
            <div class=\\"help\\">
              <div class=\\"icon-help\\">?</div>
              use <strong>lots</strong> of these
            </div>
          </span>
        <div class=\\"field required\\">
            <label class=\\"field__label\\" for=\\"the-prefix-heading_text\\">Heading text</label>
            <div class=\\"field char_field widget-text_input fieldname-heading_text\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-heading_text\\" id=\\"the-prefix-heading_text\\">Heading widget</p>
            <span></span>
          </div>
        </div>
      </div>
          </div><div class=\\"field \\">
            <label class=\\"field__label\\" for=\\"the-prefix-size\\">Size</label>
            <div class=\\"field choice_field widget-select fieldname-size\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-size\\" id=\\"the-prefix-size\\">Size widget</p>
            <span></span>
          </div>
        </div>
      </div>
          </div></div>"
`;

exports[`telepath: wagtail.blocks.StructBlock with formTemplate it renders correctly 1`] = `
"<div class=\\"custom-form-template\\">
          <p>here comes the first field:</p>
          <div class=\\"field char_field widget-text_input fieldname-heading_text\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-heading_text\\" id=\\"the-prefix-heading_text\\">Heading widget</p>
            <span></span>
          </div>
        </div>
      </div>
          <p>and here is the second:</p>
          <div class=\\"field choice_field widget-select fieldname-size\\">
        <div class=\\"field-content\\">
          <div class=\\"input\\">
            <p name=\\"the-prefix-size\\" id=\\"the-prefix-size\\">Size widget</p>
            <span></span>
          </div>
        </div>
      </div>
        </div>"
`;
